### PR TITLE
feat(java): native streaming mode deserialization

### LINKED ISSUE #1435: Fury fails when deserializing from InputStream
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-fury/issues) and found no similar issues.

### Version

0.5.0-SNAPSHOT, 0.4.1

### Component(s)

Java

### Minimal reproduce step

```java
import org.apache.fury.Fury;
import org.apache.fury.config.Language;

import java.io.BufferedInputStream;
import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.util.Arrays;

public class Reproducer {
    public static void main(String[] args) {
        Fury fury = Fury.builder()
                .withLanguage(Language.JAVA)
                .requireClassRegistration(false)
                .withMetaContextShare(false)
                .build();

        byte[] data = fury.serializeJavaObject(new TestRecord("Some string", 123));

        try {
            TestRecord result = fury.deserializeJavaObject(Arrays.copyOf(data, data.length), TestRecord.class);
            System.out.println("Byte array result: " + result);
        } catch (Exception e) {
            System.out.println("Byte array result: " + e.getClass().getSimpleName());
            e.printStackTrace();
        }

        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(Arrays.copyOf(data, data.length));
        try {
            TestRecord result = fury.deserializeJavaObject(byteArrayInputStream, TestRecord.class);
            System.out.println("ByteArrayInputStream result: " + result);
        } catch (Exception e) {
            System.out.println("ByteArrayInputStream result: " + e.getClass().getSimpleName());
            e.printStackTrace();
        }

        // hack: Fury have an optimisation for ByteArrayInputStream, which we want to avoid
        InputStream inputStream = new BufferedInputStream(new ByteArrayInputStream(Arrays.copyOf(data, data.length)));
        try {
            TestRecord result = fury.deserializeJavaObject(inputStream, TestRecord.class);
            System.out.println("InputStream result: " + result);
        } catch (Exception e) {
            System.out.println("InputStream result: " + e.getClass().getSimpleName());
            e.printStackTrace();
        }
    }

    public static class TestRecord {
        public final String a;
        public final int b;

        public TestRecord(String a, int b) {
            this.a = a;
            this.b = b;
        }

        @Override
        public String toString() {
            return "TestRecord{" +
                    "a='" + a + '\'' +
                    ", b=" + b +
                    '}';
        }
    }
}

```

### What did you expect to see?

No exception to be throwed, the data should be deserialized correctly.

### What did you see instead?

`java.lang.UnsupportedOperationException: Unsupported coder 111` when deserializing with `ByteArrayInputStream` and `java.lang.IllegalArgumentException` when deserializing with any other `InputStream`.

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by chaokunyang**
## What this PR do?
This PR implements streaming mode deserialization, which make the protocol between `Fury#serialize(buffer, o)` and `Fury#serialize(stream, o)` consistent, and make the deserialization support real streaming mode to reduce latency, and facilitate the pipeline between deserialization and transfer.

The changes inlucde:
- Introduce a `FuryStreamReader` to allow buffered streaming reading.
- Make `MemoryBuffer` support  streaming reading without introduce cost on critical path.
- Implements `InputStream` based `FuryStreamReader` for Fury.

`java.nio.Channel` based zero-copy is not implemented in this PR.

## Related issues
Closes #1435

## Media Benchmark
with this PR:
```
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  MEDIA_CONTENT         false  thrpt   50  2867721.514 ± 55102.830  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  MEDIA_CONTENT          true  thrpt   50  1841586.657 ± 46192.190  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  MEDIA_CONTENT         false  thrpt   50  2542987.512 ± 49835.129  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  MEDIA_CONTENT          true  thrpt   50  1712919.362 ± 26500.068  ops/s
```
Before this PR: 4c60eb9780ec51ffe1aac3ee2da594c0762799e0
```
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  MEDIA_CONTENT         false  thrpt   50  2971081.559 ± 52348.975  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  MEDIA_CONTENT          true  thrpt   50  1938252.031 ± 25077.333  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  MEDIA_CONTENT         false  thrpt   50  2511729.964 ± 39278.437  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  MEDIA_CONTENT          true  thrpt   50  1714610.534 ± 29778.339  ops/s
```

Baseline(7bb21d06):
```
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  MEDIA_CONTENT         false  thrpt   50  2928012.663 ± 65849.325  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  MEDIA_CONTENT          true  thrpt   50  1937231.421 ± 24616.628  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  MEDIA_CONTENT         false  thrpt   50  2539983.720 ± 43379.751  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  MEDIA_CONTENT          true  thrpt   50  1711973.100 ± 19195.431  ops/s
```

## Struct Benchmark
This PR
```
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                false  thrpt   50  2603998.829 ± 35062.683  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                 true  thrpt   50  3438343.670 ± 48911.824  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                false  thrpt   50  2631586.246 ± 34760.277  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                 true  thrpt   50  3425883.355 ± 40436.642  ops/s
```

Before This PR：
```
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                false  thrpt   50  2589559.872 ± 28062.176  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                 true  thrpt   50  3468978.381 ± 37684.130  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                false  thrpt   50  2640599.313 ± 36759.090  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                 true  thrpt   50  3451992.441 ± 51964.331  ops/s
```

Baseline：
```
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                false  thrpt   50  1576536.898 ± 54571.741  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                 true  thrpt   50  3194178.195 ± 60680.268  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                false  thrpt   50  1641797.023 ± 16222.155  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                 true  thrpt   50  3163612.265 ± 49756.995  ops/s
```

### CODE REVIEW

**Comment by chaokunyang** [comment]
Hi @Munoon, I implemented streaming deserialization for input stream. Maybe it can be taken as an example for java.nio.Channel. The implementation is a littl tricky, maybe there is a better way.

**Comment by Munoon** [comment]
> Hi @Munoon, I implemented streaming deserialization for input stream. Maybe it can be taken as an example for java.nio.Channel. The implementation is a littl tricky, maybe there is a better way.

Hi, @chaokunyang 
I feel like I'm ready with my PR, but I'm not sure how to send it, since it is built on top of your PR. I think that I will just wait until this PR will be merged. Let me know if you have a better idea.

Also, some methods of my implementation are fully copied from your `FuryInputStream` and some are partially. I think those can be extracted in the abstract class, but I'll leave it as is to make a PR simpler. Maybe it worth making a refactoring in a separate PR.

**Comment by Munoon** [comment]
Oh, and by the way, I've also received an exception when testing my channel in `MemoryBuffer#increaseReaderIndex` method. I guess, current implementation in this PR is incorrect, since the `MemoryBuffer#size` property didn't represent actual size of this buffer no more.

I've fixed this method as follow:
```java
  public void increaseReaderIndex(int diff) {
    int readerIdx = readerIndex + diff;
    if (readerIdx < 0) {
      throw new IndexOutOfBoundsException(
          String.format(
              "readerIndex: %d (expected: 0 <= readerIndex <= size(%d))", readerIdx, size));
    } else if (readerIdx > size) {
      int readCount = streamReader.fillBuffer(diff);
      if (diff > readCount) {
        throw new IndexOutOfBoundsException(
                String.format(
                        "readerIndex: %d (expected: 0 <= readerIndex <= size(%d))", readerIdx, size));
      }
    }
    this.readerIndex = readerIdx;
  }
```

**Comment by chaokunyang** [comment]
> ```java
>   public void increaseReaderIndex(int diff) {
>     int readerIdx = readerIndex + diff;
>     if (readerIdx < 0) {
>       throw new IndexOutOfBoundsException(
>           String.format(
>               "readerIndex: %d (expected: 0 <= readerIndex <= size(%d))", readerIdx, size));
>     } else if (readerIdx > size) {
>       int readCount = streamReader.fillBuffer(diff);
>       if (diff > readCount) {
>         throw new IndexOutOfBoundsException(
>                 String.format(
>                         "readerIndex: %d (expected: 0 <= readerIndex <= size(%d))", readerIdx, size));
>       }
>     }
>     this.readerIndex = readerIdx;
>   }
> ```

Nice catch, I fixed it, but fill buffer by `streamReader.fillBuffer(readerIdx - size);`

**Comment by chaokunyang** [comment]
Hi @Munoon @LiangliangSui @knutwannheden @PragmaTwice This PR is ready to review, could you help review it?

**Comment by chaokunyang** [comment]
I think it's important to our users, so I plan to include this PR in our 0.5.0 release. The original streaming is error-prone when using, and not a real streaming implementation.

**Comment by chaokunyang** [comment]
> > Hi @Munoon, I implemented streaming deserialization for input stream. Maybe it can be taken as an example for java.nio.Channel. The implementation is a littl tricky, maybe there is a better way.
> 
> Hi, @chaokunyang I feel like I'm ready with my PR, but I'm not sure how to send it, since it is built on top of your PR. I think that I will just wait until this PR will be merged. Let me know if you have a better idea.
> 
> Also, some methods of my implementation are fully copied from your `FuryInputStream` and some are partially. I think those can be extracted in the abstract class, but I'll leave it as is to make a PR simpler. Maybe it worth making a refactoring in a separate PR.

Yep, we can wait this PR merged, than you can filing a new PR

File: java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java
Comment (Munoon) [inline java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:92]: Since the `size` field did not represent actual buffer _size_, maybe it worth renaming it? WDYT @chaokunyang?

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:92]: Do you have any suggestions? I think size is still fine, bytearrayOutputStream use count, ByteBuffer use capacity

**Review by theweipeng — APPROVED** [review]

File: java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java
Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:92]: Make sense, I''ll update it later

**Comment by chaokunyang** [comment]
This PR introduce 25% performance degeneration, we can''t merge it

**Comment by chaokunyang** [comment]
> This PR introduce 25% performance degeneration, we can''t merge it

I managed to make it faster than version without streaming reading. I'll split those optimization into other PR, then we can merge this PR later.

**Comment by chaokunyang** [comment]
New readBytesString are too large for inline:
image

Old version:
image

File: java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java
Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:94]: Change this make this PR a little messy, how about renaming it in another PR.

**Comment by chaokunyang** [comment]
Hi @Munoon @LiangliangSui @theweipeng @PragmaTwice , this pr is ready for review, please take a look at it.

File: java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java
Comment (Munoon) [inline java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:2620]: What do you mean by "code size"? Bytecode size?
I've never seen such optimisation before, looks interesting. Why is it important? This also increases call stack, are you sure that it worths it?

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:2620]: Yes, it's the bytecode size of the method.  Invokng directlt will have bigger code size. Since this invocation is not  in critical path, we can move it to a separate method.

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:2620]: I'll post a benchmark later for this optimization.

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:2620]: I run a benchmark, the optimization doesn't make a big difference. I will revert it.
```
Baseline：
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                false  thrpt   50  1576536.898 ± 54571.741  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                 true  thrpt   50  3194178.195 ± 60680.268  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                false  thrpt   50  1641797.023 ± 16222.155  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                 true  thrpt   50  3163612.265 ± 49756.995  ops/s

Before This PR：
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                false  thrpt   50  2589559.872 ± 28062.176  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                 true  thrpt   50  3468978.381 ± 37684.130  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                false  thrpt   50  2640599.313 ± 36759.090  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                 true  thrpt   50  3451992.441 ± 51964.331  ops/s

This PR with fillBuffer optimization
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                false  thrpt   50  2637649.628 ± 29144.769  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                 true  thrpt   50  3481967.819 ± 38710.667  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                false  thrpt   50  2593521.429 ± 38117.758  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                 true  thrpt   50  3424332.128 ± 53014.807  ops/s

This PR without fillBuffer optimization
Benchmark                                  (bufferType)   (objectType)  (references)   Mode  Cnt        Score       Error  Units
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                false  thrpt   50  2603998.829 ± 35062.683  ops/s
UserTypeDeserializeSuite.fury_deserialize         array  STRUCT                 true  thrpt   50  3438343.670 ± 48911.824  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                false  thrpt   50  2631586.246 ± 34760.277  ops/s
UserTypeDeserializeSuite.fury_deserialize  directBuffer  STRUCT                 true  thrpt   50  3425883.355 ± 40436.642  ops/s
```

**Comment by chaokunyang** [comment]
@theweipeng @Munoon @PragmaTwice @LiangliangSui If there are no further concerns, I will merge this PR by the end of this day

File: java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java
Comment (Munoon) [inline java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:130]: Proposal: maybe instead of `BoundChecker` class we can make `MemoryBuffer` class itself extending `AbstractStreamReader`? In this case we can avoid new class initialization by simply passing `this` to the `streamReader` field.

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:130]: Could you elaborate more how `MemoryBuffer` read from the `InputStream`/`Channel`? There still a interface for such cases.

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:130]: Note: we can't make `MemoryBuffer` as a polymorphic class, it will make a large degeneration to performance

Comment (Munoon) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:130]: Yeah, sound reasonable. I just trying to find a way to avoid unnecessary object allocation, but can't find any better solution...

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/memory/MemoryBuffer.java:130]: I tried make boundchecker as a singleton, but then is can't access the index in memory buffer. Since memory buffer won't be created frequently. The object creation cost should be acceptable

**Pull request merged**
